src/imitation/util/util.py

```diff
@@ -359,3 +359,34 @@ def get_first_iter_element(iterable: Iterable[T]) -> Tuple[T, Iterable[T]]:
         return_iterable = iterable
 
     return first_element, return_iterable
+
+
+def compute_state_entropy(
+    obs: th.Tensor,
+    all_obs: th.Tensor,
+    k: int,
+) -> th.Tensor:
+    """Compute the state entropy given by KNN distance.
+
+    Args:
+        obs: A batch of observations.
+        all_obs: The tensor of all states to compare to.
+        k: the number of neighbors to consider
+
+    Returns:
+        A tensor containing the state entropy for `obs`.
+    """
+    assert obs.shape[1:] == all_obs.shape[1:]
+    with th.no_grad():
+        non_batch_dimensions = tuple(range(2, len(obs.shape) + 1))
+        distances_tensor = th.linalg.vector_norm(
```

Comment:
This tensor could get quite big O(batch_size * total_timesteps * unsupervised_pretraining_frac). Don't want to premature optimize, but we should do some testing before merging to see if this is a problem in practice, I'm sure there's some more efficient ways to find KNN distances.

Reply:
Yeah, the original implementation actually does this by parts: https://github.com/rll-research/BPref/blob/f3ece2ecf04b5d11b276d9bbb19b8004c29429d1/agent/sac.py#L13

Shall we implement something similar?

Reply:
Ok, I went ahead and pushed the batched version.